### PR TITLE
dockerversion placeholder for library-import

### PR DESCRIPTION
**Pull request by vdemeester**
Little bit scared to do this PR 😱 but here we go. This is _implementing_ the idea of @erikh from #11699. Kinda want **some thoughts and opinions** on this and ideas to move forward, that's why the PR 🎅.

> what about doing this, instead, as a placeholder? https://gist.github.com/erikh/e6f0c5e4b7b6c2483d52
> 
> you could leave this in the repository so it's still go-gettable, but then overwrite it during the build. how does this sound?

To add to the discussion, quotation from @ibuildthecloud 

> I know people shouldn't import non-pkg code but I don't like that effectively docker makes it not possible. If people want to import stuff at their own risk, just let them.

/ping @tianon @jfrazelle @docker/maintainers 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by erikh** [comment]
needs moar golint I think. thanks for doing this!

**Comment by vdemeester** [comment]
oh.. that ain't good 😱, didn't though of that :smile_cat:

File: autogen/dockerversion/dockerversion.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// Package dockerversion is auto-generated at build-time
+package dockerversion
+
+// Default build-time variable for library-import.
+// This file is overridden on build with build-time informations.
+const (
+	GITCOMMIT string = "library-import"
```
Comment (stevvooe) [inline autogen/dockerversion/dockerversion.go]: These need to follow the Go style guidelines. `GitCommit`, `Version`, etc.

File: autogen/dockerversion/dockerversion.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Package dockerversion is auto-generated at build-time
```
Comment (stevvooe) [inline autogen/dockerversion/dockerversion.go]: No reason to hide file in "autogen" package. It can just be under "dockerversion" or "version".

**Comment by stevvooe** [comment]
@vdemeester This is awesome. Mostly just detail cleanup needed.

Also, I would suggest generating the file with `go generate`.

> I know people shouldn't import non-pkg code but I don't like that effectively docker makes it not possible. If people want to import stuff at their own risk, just let them.

This is a silly restriction and reduces code sharing. Packages are packages. If you export a symbol, assume the package will be imported. If you don't want it imported, don't export the symbols. Go also has the "internal" package structure to disallow cross-project import, if needed.

**Comment by tianon** [comment]
I'm not opposed to something like this, but I think we definitely need to make sure that we don't have our build process modifying Git-tracked files unless it can revert the changes it makes.  I think we can combat this by using build tags -- ie, have `hack/make.sh` include a build tag of `autogen` or something like that, and have the checked in file ignored if that tag is specified, but the actual autogenerated file is only used with the tag.  That way, I think we'd get the best of both worlds as much as possible.

**Comment by erikh** [comment]
Seems legit as far as solutions go.

Glad to see this is moving forward. Thanks all.

File: autogen/dockerversion/dockerversion.go (hunk 1 above)
Comment (vdemeester) [inline autogen/dockerversion/dockerversion.go]: Didn't want to make too much changes across the sources, that's why I kept these `CAPS` and used `const` (that way golint is happy).

I can make the change but I felt the change should be not too much intrusive (at first :stuck_out_tongue_closed_eyes:)

**Comment by vdemeester** [comment]
> I'm not opposed to something like this, but I think we definitely need to make sure that we don't have our build process modifying Git-tracked files unless it can revert the changes it makes. I think we can combat this by using build tags -- ie, have hack/make.sh include a build tag of autogen or something like that, and have the checked in file ignored if that tag is specified, but the actual autogenerated file is only used with the tag. That way, I think we'd get the best of both worlds as much as possible.

Oh didn't thought of that, good idea :wink:

File: autogen/dockerversion/dockerversion.go (hunk 1 above)
Comment (calavera) [inline autogen/dockerversion/dockerversion.go]: I agree with @vdemeester, that's not in the scope of this PR.

**Comment by calavera** [comment]
I agree with @stevvooe that if we're going to have this static it should not be under `autogen`. Maybe we could have it under `docker/version`, so the import path would be something like `github.com/docker/docker/version`. We could have this static file called `version_lib.go` and let make create a `version_autogen.go`.

:+1: to build tags too.

**Comment by stevvooe** [comment]
We should use [go generate](https://blog.golang.org/generate) to generate the file and the linker to modify the values to be accurate for the current state of the repository.

**Comment by tianon** [comment]
@stevvooe we used to use `-X` to set these, but that doesn't work with `gccgo`, which is why we had to introduce this nasty autogenerated code in the first place :cry:

File: hack/make.sh [hunk 3]
```diff
@@ -96,6 +96,9 @@ if [ ! "$GOPATH" ]; then
 	exit 1
 fi
 
+# FIXME Maybe move this elsewhere ?
+DOCKER_BUILDTAGS+=" autogen"
```
Comment (vdemeester) [inline hack/make.sh]: Was wondering if it was the right place to put it.

Comment (tianon) [inline hack/make.sh]: Probably down in `ORIG_BUILDFLAGS` directly where we have several more that are already statically set unconditionally.

**Comment by calavera** [comment]
nice, LGTM.

File: hack/make/.go-autogen [hunk 4]
```diff
@@ -2,12 +2,14 @@
 
 rm -rf autogen
 
-mkdir -p autogen/dockerversion
-cat > autogen/dockerversion/dockerversion.go <<DVEOF
-// AUTOGENERATED FILE; see $BASH_SOURCE
-package dockerversion
-
-var (
+cat > version/version_autogen.go <<DVEOF
```
Comment (tianon) [inline hack/make/.go-autogen]: Since this file isn't in that `autogen` directory anymore, do you think we should add it to `.gitignore` too?

**Comment by tianon** [comment]
One small question, but otherwise LGTM :+1:

File: hack/make/.go-autogen (hunk 4 above)
Comment (vdemeester) [inline hack/make/.go-autogen]: Ah yes, I did forget about updating `.gitignore` :sweat_smile:.

File: .gitignore [hunk 5]
```diff
@@ -18,6 +18,7 @@
 Vagrantfile
 a.out
 autogen/
+version/version_autogen.go
```
Comment (tianon) [inline .gitignore]: This file's sorted, isn't it? :wink:

Comment (vdemeester) [inline .gitignore]: hehe maybe :stuck_out_tongue_closed_eyes:

**Comment by tianon** [comment]
LGTM :+1:

**Comment by dmcgowan** [comment]
LGTM

**Comment by calavera** [comment]
I just tested it and the info looks as expected:

```
➜  docker git:(vdemeester-11699-auto-libraryimport) bundles/1.9.0-dev/binary/docker version
Client:
 Version:      1.9.0-dev
 API version:  1.22
 Go version:   go1.4.3
 Git commit:   d5cd032
 Built:        Tue Oct 27 20:35:12 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.22
 Go version:   go1.4.3
 Git commit:   d5cd032
 Built:        Tue Oct 27 20:35:12 UTC 2015
 OS/Arch:      linux/amd64
```

Merging.

**Comment by vdemeester** [comment]
:tada: :heart_eyes_cat: :kissing_cat:

**Comment by ibuildthecloud** [comment]
This is beautiful.  Thanks!!! @vdemeester.  Time to update Godeps in libcompose :)

**Comment by thaJeztah** [comment]
Cool, thanks @erikh and @vdemeester !

File: version/version_lib.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// +build !autogen
+
+// Package version is auto-generated at build-time
+package version
+
+// Default build-time variable for library-import.
+// This file is overridden on build with build-time informations.
+const (
+	GITCOMMIT string = "library-import"
```
Comment (stevvooe) [inline version/version_lib.go]: How did this get merged without following the coding standards?

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#coding-style

Comment (tianon) [inline version/version_lib.go]: I'm pretty sure this specific constant predates the coding style, but a PR
to update this to conform would definitely be great IMO.

Comment (vdemeester) [inline version/version_lib.go]: @stevvooe I guess this is based https://github.com/docker/docker/pull/17350#discussion_r43043783 and the fact that the generated file were already like that (so didn't changed it). And `golint` is actually *OK* with ALL_CAPS constants.

Comment (vdemeester) [inline version/version_lib.go]: @tianon @stevvooe I can do a quick follow-up, more than fine with it :wink:

Comment (stevvooe) [inline version/version_lib.go]: @vdemeester I was under the impression these were new. I suppose we can keep them around and update them in another PR. I'm surprised `golint` let this pass.

Comment (vdemeester) [inline version/version_lib.go]: @stevvooe #17420 :wink:

**Pull request merged**